frame/elections-phragmen/src/lib.rs

```diff
@@ -483,17 +488,11 @@ decl_module! {
 
 			let valid = Self::is_defunct_voter(&votes);
 			if valid {
-				// reporter will get the voting bond of the target
-				T::Currency::repatriate_reserved(&target, &reporter, T::VotingBond::get(), BalanceStatus::Free)?;
-				// remove the target. They are defunct.
-				Self::do_remove_voter(&target, false);
-			} else {
-				// slash the bond of the reporter.
-				let imbalance = T::Currency::slash_reserved(&reporter, T::VotingBond::get()).0;
-				T::BadReport::on_unbalanced(imbalance);
-				// remove the reporter.
-				Self::do_remove_voter(&reporter, false);
+				// remove the defunct target while unreserving their bond.
+				Self::do_remove_voter(&target, true);
```

Comment:
So maybe we can do some kind of refund here, but I think that would open a new attack vector: 

The proxy can set a wrong vote, report it, and it will be for free. 

All in all, it is bit unfortunate that all modules need now be proxy-aware. The pallet was safe for a runtime without proxy and now it is not. Maybe we can come up with different and better ways to deal with such issues.

Reply:
> The proxy can set a wrong vote

No - voting should require a deposit on-chain equal to the resources that the vote takes. If the proxy sets a wrong vote, they can tie up funds from the main account into a deposit, but that's fine as the main account can remove the proxy and then free up the vote.

We don't have a concept of "defunct voter" any more. All voting is paid for and is allowed to remain on-chain until the voter (or their proxy) chooses to remove it, even if it is "defunct" (i.e. doesn't actually contributes toward any tallies).

The weight associated with this function should be done normally (i.e. it's enough to ensure that if it's called incorrectly (and results in a no-op), then the computation is paid for).

Reply:
Yes I first recommended that maybe this _cleanup_ operation can be made a free transaction if it was successful because it is doing the chain some favour, but as you said given the vote this is not really a favour anymore, thus okay as-is.  

> Crucially, we need to change vote from a fixed deposit to one that properly recognises the costs associated with the vote.

Addressed. 

> This is wrong - they don't.

Well, I can't see how we could have had the same vulnerability without the concept of proxies. 

For example, I was thinking: Ideally we could allow the proxy module to configure each proxy with some call indices that can be allowed (instead of the current `enum ProxyType` which is more high level and only distinguishes certain operations). Then, the person who is delegating a proxy has fine grained control over which calls their proxy can dispatch and can (nay -- _should_) prohibit the proxy from calling into such functions that can drain their bond. 

In that case, if they happen to allow it, it is no longer our responsibility and they've messed up their proxy configurations.

Reply:
Well, I would say that it didn't really strictly adhere to game-theoretic principles that we eventually developed elsewhere in the codebase (i.e. notably that any additional state should require a deposit according to the additional storage/computation it implies), and this was done even before we had this particular attack vector.

If you ensure that the actor themself always pays for their actions, then you're in good territory. The issue here was that "the chain" kind of paid for the repurcussions of the `vote` action as evidenced by the chain being willing to slash&repatriate funds in order to incentivize its cleanup. This should not be necessary as long as the voter is paying their way properly with a deposit.

In general, we should be very careful about repatriating funds. In hindsight, this is probably quite a big footgun and we might even want to deprecate it entirely.